File: src/flyarena/research.py
"""Provider-independent research contracts and same-condition comparisons."""
from __future__ import annotations

from typing import Literal
import math
import numpy as np
from pydantic import Field, StrictInt, model_validator
from .common import digest
from .contracts import StrictModel, InterventionSpec


class BackendProfile(StrictModel):
    id: str
    backend_id: str
    model_id: str
    sensor_id: str
    readout_id: str
    motor_id: str | None = None
    protocol_id: str | None = None
    actual_backend: str | None = None
    embodiment_id: str
    ready: bool
    hashes: dict[str, str]
    capabilities: list[str] | dict[str, bool | str]
    limitations: list[str] = Field(default_factory=list)
    unavailable: list[str] = Field(default_factory=list)
    quality_gates: dict[str, bool] = Field(default_factory=dict)
    reason: str | None = None


class ScenarioSpec(StrictModel):
    schema_version: Literal['scenario/v1'] = 'scenario/v1'
    geometry_id: str
    stimulus_id: str
    task_id: str
    probe_id: str
    seed: int = Field(ge=0, le=2**31-1)


class ExperimentSpec(StrictModel):
    fly_id: str = Field(pattern=r'^[0-9a-f]{32}$')
    probe_id: str = Field(default='gradient-v2', min_length=1, max_length=80)
    seeds: list[StrictInt] = Field(default_factory=lambda: [42], min_length=1, max_length=8)
    duration_seconds: int = Field(default=3, ge=1, le=30, strict=True)

    @model_validator(mode='after')
    def distinct_seeds(self):
        if len(set(self.seeds)) != len(self.seeds) or any(type(s) is not int or s < 0 or s > 2**31-1 for s in self.seeds):
            raise ValueError('Seeds must be distinct nonnegative int32 values')
        return self


class ConditionSpec(StrictModel):
    """Full frozen conditions; deliberately contains no subject or provenance."""
    schema_version: Literal['condition/v1'] = 'condition/v1'
    probe: dict
    profile: dict
    runtime: dict
    scene: dict
    seed: int = Field(ge=0, le=2**31-1)
    duration_seconds: int = Field(ge=1, le=30)


def condition_key(condition: ConditionSpec) -> str:
    return digest(condition.model_dump(mode='json'))


def scientific_conditions(condition: ConditionSpec) -> dict:
    return {'profile':condition.profile,'runtime':condition.runtime,'scene':condition.scene,
            'seed':condition.seed,'duration_seconds':condition.duration_seconds,'probe_id':condition.probe['id'],
            'ablation':None,'stimulus':'normal','decoder_version':'v2'}


def run_key(condition: ConditionSpec | str, artifact_id: str) -> str:
    if len(artifact_id) != 64 or any(c not in '0123456789abcdef' for c in artifact_id):
        raise ValueError('Invalid artifact digest')
    key = condition_key(condition) if isinstance(condition, ConditionSpec) else condition
    if len(key) != 64 or any(c not in '0123456789abcdef' for c in key):
        raise ValueError('Invalid condition digest')
    return digest({'condition_key': key,
                   'artifact_id': artifact_id})


class TrajectoryPoint(StrictModel):
    time: float = Field(ge=0)
    x: float
    y: float
    yaw: float


class SceneObstacle(StrictModel):
    position: list[float] = Field(min_length=3, max_length=3)
    size: list[float] = Field(min_length=3, max_length=3)


class SceneFood(StrictModel):
    id: str
    position: list[float] = Field(min_length=3, max_length=3)
    initial: float = Field(ge=0)


class ProbeScene(StrictModel):
    id: str
    size: float = Field(gt=0)
    spawns: list[list[float]]
    obstacles: list[SceneObstacle]
    food: list[SceneFood] = Field(min_length=1)
    mirror: Literal[-1, 1]
    odor_sigma_xy_mm: list[float] | None = Field(default=None, min_length=2, max_length=2)
    chemical_policy: str
    cue_on_seconds: float = Field(ge=0)
    cue_off_seconds: float | None

    @model_validator(mode='after')
    def valid_geometry(self):
        if self.odor_sigma_xy_mm is not None and min(self.odor_sigma_xy_mm) <= 0:
            raise ValueError('Chemical spread must be positive')
        if not self.spawns or any(len(s) != 3 or not all(math.isfinite(x) for x in s) for s in self.spawns):
            raise ValueError('Scene requires finite x/y/yaw spawns')
        if len({f.id for f in self.food}) != len(self.food) or any(min(o.size) <= 0 for o in self.obstacles):
            raise ValueError('Invalid scene food identities or obstacle size')
        return self


class PhenotypeReport(StrictModel):
    schema_version: Literal['probe-report/v2']
    fly_id: str = Field(pattern=r'^[0-9a-f]{32}$')
    artifact_id: str = Field(pattern=r'^[0-9a-f]{64}$')
    condition_key: str = Field(pattern=r'^[0-9a-f]{64}$')
    probe_id: str
    seed: int = Field(ge=0, le=2**31-1)
    duration_seconds: int = Field(ge=1, le=30)
    status: Literal['complete', 'partial', 'failed', 'error']
    trajectory: list[TrajectoryPoint]
    metrics: dict[str, float | None]
    neural_trace: list[dict[str, float | None]]
    receipt_sha256: str | None = Field(default=None, pattern=r'^[0-9a-f]{64}$')
    profile: dict
    scene: ProbeScene
    frames: list | dict | None = None
    error: str | None = None

    @model_validator(mode='after')
    def time_order(self):
        times = [p.time for p in self.trajectory]
        if any(b <= a for a, b in zip(times, times[1:])):
            raise ValueError('Trajectory times must be strictly increasing')
        if times and times[-1] > self.duration_seconds + 1e-6:
            raise ValueError('Trajectory exceeds condition horizon')
        if self.status == 'complete' and (not self.receipt_sha256 or len(times) < 2):
            raise ValueError('Complete report requires receipt and trajectory')
        if self.status == 'complete' and (abs(times[0]) > 1e-6 or abs(times[-1]-self.duration_seconds) > 1e-6):
            raise ValueError('Complete trajectory must cover the entire condition horizon')
        BackendProfile.model_validate(self.profile)
        for point in self.neural_trace:
            if any(v is not None and not math.isfinite(v) for v in point.values()):
                raise ValueError('Nonfinite neural trace')
        return self


class PairedComparison(StrictModel):
    seed: int
    fly_id: str
    reference_fly_id: str
    deltas: dict[str, float | None]
    trajectory_divergence_mm: float | None


class ComparisonReport(StrictModel):
    schema_version: Literal['comparison/v1'] = 'comparison/v1'
    paired: list[PairedComparison]
    conditions: list[str]


class SeasonSpec(StrictModel):
    id: str
    ranking_policy: Literal['ranking/v1'] = 'ranking/v1'
    runtime_sha256: str
    scenario_id: str
    mode: Literal['forage', 'contest', 'sumo']


def compare_reports(reports: list[PhenotypeReport | dict], subjects: list[dict] | None = None) -> ComparisonReport:
    parsed = [r if isinstance(r, PhenotypeReport) else PhenotypeReport.model_validate(r) for r in reports]
    if not parsed:
        raise ValueError('Comparison requires reports')
    if any(r.status != 'complete' for r in parsed):
        raise ValueError('Comparison requires complete verified trials; partial/failed is not zero')
    ids = [s['fly_id'] for s in subjects] if subjects else list(dict.fromkeys(r.fly_id for r in parsed))
    if len(ids) != 3 or len(set(ids)) != 3:
        raise ValueError('Comparison requires independent wildtype, official and design subjects')
    paired, conditions = [], []
    for seed in sorted({r.seed for r in parsed}):
        group = [r for r in parsed if r.seed == seed]
        if len(group) != 3 or {r.fly_id for r in group} != set(ids):
            raise ValueError('Missing or duplicate subject trial')
        if len({r.condition_key for r in group}) != 1 or len({(r.probe_id, r.duration_seconds, digest(r.profile)) for r in group}) != 1:
            raise ValueError('Mismatched frozen conditions')
        conditions.append(group[0].condition_key)
        by_id = {r.fly_id: r for r in group}
        ref = by_id[ids[0]]
        times = np.array([p.time for p in ref.trajectory])
        for ident in ids[1:]:
            trial = by_id[ident]
            tt = np.array([p.time for p in trial.trajectory])
            if len(tt) != len(times) or not np.allclose(tt,times,atol=1e-10,rtol=0):
                raise ValueError('Mismatched trajectory time grids or horizons')
            grid = np.unique(np.concatenate([times, tt]))
            def xy(r):
                t = [p.time for p in r.trajectory]
                return np.column_stack([np.interp(grid, t, [getattr(p, axis) for p in r.trajectory]) for axis in ('x', 'y')])
            distances = np.linalg.norm(xy(trial)-xy(ref), axis=1)
            divergence = float(np.trapezoid(distances, grid)/(grid[-1]-grid[0]))
            deltas = {k: (trial.metrics[k]-ref.metrics[k] if trial.metrics.get(k) is not None and ref.metrics.get(k) is not None else None)
                      for k in sorted(trial.metrics.keys() | ref.metrics.keys())}
            paired.append(PairedComparison(seed=seed, fly_id=ident, reference_fly_id=ref.fly_id,
                                           deltas=deltas, trajectory_divergence_mm=divergence))
    return ComparisonReport(paired=paired, conditions=conditions)

File: web/src/shared/research.ts
import type {Fly} from '../types'
export type CellSelector={class?:string;type?:string;side?:'L'|'R';ids?:string[]}
export type InterventionSpec={selector:{pre?:CellSelector;post?:CellSelector};scale:number}
export type SubjectRole='wildtype'|'official'|'design'
export type ExperimentSubject={role:SubjectRole;fly_id:string;name:string;artifact_id:string;parent_id?:string|null;reference_kind?:Fly['reference_kind'];submission_channel?:Fly['submission_channel'];release_id?:string|null}
export type ScenarioSpec={id?:string;geometry?:{size?:number;bounds?:number[];obstacles?:{position:number[];size:number[]}[]};stimulus?:unknown;task?:unknown;probe?:unknown;[key:string]:unknown}
export type BackendProfile={id:string;backend_id:string;model_id:string;sensor_id:string;readout_id:string;embodiment_id:string;ready:boolean;reason?:string;reduced_mode?:boolean;limitations?:string[];unavailable?:string[];motor_id?:string;protocol_id?:string;hashes:Record<string,string>;capabilities:Record<string,unknown>|string[]}
export type Probe={id:string;name:string;description:string;capabilities:Record<string,unknown>|string[];scenario?:ScenarioSpec;scenario_spec?:ScenarioSpec;[key:string]:unknown}
export type ResearchCatalog={schema_version:string;probes:Probe[];profile:BackendProfile;references:{wildtype:Fly|null;official:Fly|null};backends:unknown[];integrations:unknown[]}
export type ExperimentSpec={fly_id:string;probe_id:string;seeds:number[];duration_seconds:number}
export type TrajectoryPoint={time:number;x:number;y:number;yaw:number}
export type NeuralPoint={time:number;drive_left?:number;drive_right?:number;[key:string]:unknown}
export type ProbeReport={schema_version:string;fly_id:string;artifact_id:string;condition_key:string;probe_id:string;seed:number;duration_seconds:number;status:string;trajectory:TrajectoryPoint[];metrics:Record<string,number|null>;neural_trace:NeuralPoint[];receipt_sha256:string;profile:BackendProfile;scenario?:ScenarioSpec;geometry?:ScenarioSpec['geometry'];scene?:ResearchScene;error?:string|null;[key:string]:unknown}
export type ComparisonReport={schema_version:string;paired:{seed:number;fly_id:string;reference_fly_id:string;deltas:Record<string,number|null>;trajectory_divergence_mm:number|null}[];conditions:unknown[];summary?:unknown}
export type Experiment={id:string;owner:string;status:'queued'|'running'|'complete'|'failed';error:string|null;spec:ExperimentSpec;subjects:ExperimentSubject[];reports:ProbeReport[];comparison:ComparisonReport|null;created:number}
export const roleStyles:Record<SubjectRole,{color:string;dash:string;symbol:string}>={wildtype:{color:'#ffffff',dash:'',symbol:'○'},official:{color:'#e8b45d',dash:'9 5',symbol:'△'},design:{color:'#8ee0ba',dash:'3 4',symbol:'◇'}}
export function finite(value:unknown):value is number{return typeof value==='number'&&Number.isFinite(value)}
export function metricUnit(key:string){if(/(?:censored|choice|success|count|intake|energy)/.test(key))return '—';if(/(?:_mm_s|speed)/.test(key))return 'mm/s';if(/(?:_rad_s|angular_velocity)/.test(key))return 'rad/s';if(/(?:_rad|turning)/.test(key))return 'rad';if(/(?:_mm|path_length|distance|divergence)/.test(key))return 'mm';if(/(?:_seconds|_s$|latency|contact_time)/.test(key))return 's';if(/(?:_hz|rate)/.test(key))return 'Hz';return '—'}
export function validateHorizon(value:number){if(!Number.isInteger(value)||value<1||value>30)throw new Error('Horizon must be an integer from 1 to 30 seconds.');return value}
export function designRoleLabel(owner:string|undefined,viewer:string|undefined){return viewer&&owner===viewer?'Your design':'Submitted design'}
export type ResearchScene={size:number;spawns?:number[][];obstacles:{position:number[];size:number[]}[];food:{id:string;position:number[];initial:number}[]}
export function plotTransform(points:{x:number;y:number}[],scene?:ResearchScene){
 const half=scene?.size?scene.size/2:0;const xs=points.map(p=>p.x),ys=points.map(p=>p.y);
 for(const f of scene?.food||[]){xs.push(f.position[0]);ys.push(f.position[1])}
 for(const o of scene?.obstacles||[]){xs.push(o.position[0]-o.size[0]/2,o.position[0]+o.size[0]/2);ys.push(o.position[1]-o.size[1]/2,o.position[1]+o.size[1]/2)}
 const x0=Math.min(...xs,half?-half:Infinity)-1,x1=Math.max(...xs,half||-Infinity)+1,y0=Math.min(...ys,half?-half:Infinity)-1,y1=Math.max(...ys,half||-Infinity)+1;
 const scale=Math.min(620/Math.max(x1-x0,1),350/Math.max(y1-y0,1));
 return {x0,x1,y0,y1,scale,sx:(x:number)=>360+(x-(x0+x1)/2)*scale,sy:(y:number)=>220-(y-(y0+y1)/2)*scale}
}
export function parseSeeds(value:string){const tokens=value.split(/[,\s]+/).filter(Boolean);if(!tokens.length||tokens.length>8)throw new Error('Enter between 1 and 8 distinct seeds.');const seeds=tokens.map(Number);if(seeds.some(n=>!Number.isSafeInteger(n)||n<0||n>2147483647))throw new Error('Seeds must be integers from 0 to 2147483647.');if(new Set(seeds).size!==seeds.length)throw new Error('Seeds must be distinct.');return seeds}
/** Refuse alignment claims if any frozen subject is absent or the reported grids differ. */
export function alignment(reports:ProbeReport[],subjects:ExperimentSubject[]){if(!subjects.length||subjects.some(s=>!reports.some(r=>r.fly_id===s.fly_id&&r.artifact_id===s.artifact_id)))return 'pending';if(reports.some(r=>r.status!=='complete'||!r.receipt_sha256||!r.condition_key||!r.trajectory?.length))return 'pending';if(reports.some(r=>r.trajectory.some((p,i)=>!finite(p.time)||!finite(p.x)||!finite(p.y)||!finite(p.yaw)||p.time<0||p.time>r.duration_seconds||(i>0&&p.time<=r.trajectory[i-1].time))))return 'mismatch';const first=reports[0];return reports.every(r=>r.condition_key===first.condition_key&&r.duration_seconds===first.duration_seconds&&r.trajectory.length===first.trajectory.length&&r.trajectory.every((p,i)=>finite(p.time)&&p.time===first.trajectory[i].time))?'matched':'mismatch'}
export function sampleAt(points:TrajectoryPoint[],time:number){const valid=points.filter(p=>finite(p.time)&&finite(p.x)&&finite(p.y));if(!valid.length||time<valid[0].time||time>valid[valid.length-1].time)return null;let i=valid.findIndex(p=>p.time>=time);if(i<1)return valid[0];const a=valid[i-1],b=valid[i],f=(time-a.time)/(b.time-a.time||1);return {...a,time,x:a.x+(b.x-a.x)*f,y:a.y+(b.y-a.y)*f}}
